mlir/lib/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation.cpp

```diff
@@ -856,13 +857,32 @@ FailureOr<Operation *> BufferDeallocation::handleInterface(CallOpInterface op) {
 FailureOr<Operation *>
 BufferDeallocation::handleInterface(MemoryEffectOpInterface op) {
   auto *block = op->getBlock();
+  OpBuilder builder = OpBuilder::atBlockBegin(block);
 
-  for (auto operand : llvm::make_filter_range(op->getOperands(), isMemref))
-    if (op.getEffectOnValue<MemoryEffects::Free>(operand).has_value())
-      return op->emitError(
-          "memory free side-effect on MemRef value not supported!");
+  for (auto operand : llvm::make_filter_range(op->getOperands(), isMemref)) {
+    if (op.getEffectOnValue<MemoryEffects::Free>(operand).has_value()) {
+      if (!op->hasAttr(BufferizationDialect::kManualDeallocation))
+        return op->emitError(
+            "memory free side-effect on MemRef value not supported!");
```

Comment:
I don't quite get this message here? It does not seems to relate to the conditions that lead here, or am I missing something?

Reply:
This is one of the things that confused me in #72289.

Reply:
The buffer deallocation pass fails if there are any "free" side effects (memory deallocations) in the input IR. We cannot handle such input programs at the moment.

There is one exception: Existing buffer deallocation ops are allowed if they are annotated with `bufferization.manual_deallocation`. The corresponding allocation must then also be annotated with that attribute. The buffer deallocation pass can completely ignore such buffers. (Internally, they are assigned an ownership indicator of "false".)

In essence, `bufferization.manual_deallocation` is a way exclude parts of the input IR from the buffer deallocation pass.

The code here checks if the op has a free side effect. If that is the case and there is no `manual_deallocation` attribute on the op, we reject the IR.